cli/internal/runcache/runcache.go

```diff
@@ -164,17 +170,33 @@ func (tc TaskCache) RestoreOutputs(terminal *cli.PrefixedUi, logger hclog.Logger
 		}
 		return false, nil
 	}
-	// TODO(gsoltis): check if we need to restore goes here
-	// That will be an opportunity to prune down the set of outputs as well
-	hit, _, _, err := tc.rc.cache.Fetch(tc.rc.repoRoot.ToString(), tc.hash, tc.repoRelativeGlobs)
+	changedOutputGlobs, err := tc.rc.outputWatcher.GetChangedOutputs(tc.hash, tc.repoRelativeGlobs)
 	if err != nil {
-		return false, err
-	} else if !hit {
-		if tc.rc.cacheMissLogsMode != NoLogs {
-			terminal.Output(fmt.Sprintf("cache miss, executing %s", ui.Dim(tc.hash)))
+		logger.Warn(fmt.Sprintf("Failed to check if we can skip restoring outputs for %v: %v. Proceeding to check cache", tc.pt.TaskID, err))
+		terminal.Warn(ui.Dim(fmt.Sprintf("Failed to check if we can skip restoring outputs for %v: %v. Proceeding to check cache", tc.pt.TaskID, err)))
+		changedOutputGlobs = tc.repoRelativeGlobs
+	}
+	if len(changedOutputGlobs) == 0 {
+		logger.Debug(fmt.Sprintf("Skipping cache check for %v, outputs have not changed since previous run.", tc.pt.TaskID))
+	} else {
+		// Note that we currently don't use the output globs when restoring, but we could in the
+		// future to avoid doing unnecessary file I/O
```

Comment:
We should segment cache addressing by non-colliding globs. This will likely prevent exhausting remote space and bandwidth for image files that exist in a repo, etc.

(just a random thought while reading)

Reply:
"non-colliding globs" is a somewhat difficult thing to determine in the general case, but we might be able to make some headway on it for some cases.

In this case, we don't care about whether or not they overlap, we can invalidate them independently and then only glob for the ones that have been invalidated.